File: src/CircleAI.Inference/NightlyAdapterTrainer.cs
// NightlyAdapterTrainer.cs
//
// (Phase D3) Periodically drains the FeedbackTrainingQueue, runs LoRA
// gradient steps against the current model handle, saves the adapter
// to disk, and applies it atomically. Idle-and-charging gating is
// host-supplied via the ShouldFireNow predicate.

using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace CircleAI.Inference;

/// <param name="MinBatchSize">Minimum samples to bother training. Skip otherwise.</param>
/// <param name="MaxSamplesPerRun">Cap per nightly run so a backlog can't lock the device.</param>
/// <param name="LearningRate">Adam-style LR for the LoRA adapter parameters.</param>
/// <param name="LoRARank">Rank of the LoRA decomposition; lower = smaller adapter.</param>
/// <param name="AdapterPath">Where to persist the trained adapter file.</param>
/// <param name="Interval">How often to check whether to train. Default 6 hours.</param>
/// <param name="ShouldFireNow">Optional gate (battery, charging, idle) — defaults to "always".</param>
/// <param name="Tokenizer">Tokeniser to convert text → int IDs. Falls back to char-level mapping if null.</param>
public sealed record NightlyAdapterTrainerOptions(
    int           MinBatchSize     = 16,
    int           MaxSamplesPerRun = 256,
    float         LearningRate     = 1e-4f,
    int           LoRARank         = 8,
    string        AdapterPath      = "circleai-lora.mnn",
    TimeSpan?     Interval         = null,
    Func<bool>?   ShouldFireNow    = null,
    Func<string, int[]>? Tokenizer = null);

public sealed class NightlyAdapterTrainer : IHostedService, IAsyncDisposable
{
    private readonly IFeedbackTrainingQueue _queue;
    private readonly LoRAAdapterManager _adapter;
    private readonly NightlyAdapterTrainerOptions _opts;
    private readonly ILogger<NightlyAdapterTrainer> _logger;
    private CancellationTokenSource? _cts;
    private Task? _loop;

    public NightlyAdapterTrainer(
        IFeedbackTrainingQueue queue,
        LoRAAdapterManager adapter,
        NightlyAdapterTrainerOptions opts,
        ILogger<NightlyAdapterTrainer>? logger = null)
    {
        _queue   = queue   ?? throw new ArgumentNullException(nameof(queue));
        _adapter = adapter ?? throw new ArgumentNullException(nameof(adapter));
        _opts    = opts    ?? throw new ArgumentNullException(nameof(opts));
        _logger  = (ILogger<NightlyAdapterTrainer>?)logger ?? NullLogger<NightlyAdapterTrainer>.Instance;
    }

    public Task StartAsync(CancellationToken cancellationToken)
    {
        if (_cts is not null) return Task.CompletedTask;
        _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        _loop = Task.Run(() => LoopAsync(_cts.Token), CancellationToken.None);
        _logger.LogInformation("[NightlyAdapterTrainer] started; interval={Interval}", _opts.Interval ?? TimeSpan.FromHours(6));
        return Task.CompletedTask;
    }

    public async Task StopAsync(CancellationToken cancellationToken)
    {
        if (_cts is null) return;
        _cts.Cancel();
        try { if (_loop is not null) await _loop.WaitAsync(cancellationToken).ConfigureAwait(false); }
        catch (OperationCanceledException) { }
        _cts.Dispose();
        _cts = null;
        _loop = null;
    }

    public async ValueTask DisposeAsync() => await StopAsync(CancellationToken.None).ConfigureAwait(false);

    private async Task LoopAsync(CancellationToken ct)
    {
        var interval = _opts.Interval ?? TimeSpan.FromHours(6);
        while (!ct.IsCancellationRequested)
        {
            try
            {
                if (_opts.ShouldFireNow is null || _opts.ShouldFireNow())
                    await RunOnceAsync(ct).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "[NightlyAdapterTrainer] run failed");
            }
            try { await Task.Delay(interval, ct).ConfigureAwait(false); }
            catch (OperationCanceledException) { return; }
        }
    }

    /// <summary>(Phase D3) Drain + train in one pass. Public so a host can trigger manually.</summary>
    public async ValueTask RunOnceAsync(CancellationToken ct = default)
    {
        if (_queue.Pending < _opts.MinBatchSize)
        {
            _logger.LogDebug("[NightlyAdapterTrainer] skipping; pending={Pending} < min={Min}", _queue.Pending, _opts.MinBatchSize);
            return;
        }

        var samples = await _queue.DrainAsync(_opts.MaxSamplesPerRun, ct).ConfigureAwait(false);
        if (samples.Count == 0) return;

        var tokenizer = _opts.Tokenizer ?? CharTokenizer;
        var totalLoss = 0f;
        var stepCount = 0;
        foreach (var sample in samples)
        {
            ct.ThrowIfCancellationRequested();
            try
            {
                var input  = tokenizer(sample.UserText);
                var target = tokenizer(sample.Polarity >= 0 ? sample.PreferredText : sample.AssistantText);
                if (input.Length == 0 || target.Length == 0) continue;

                var loss = _adapter.TrainStep(input, target, _opts.LearningRate, _opts.LoRARank);
                totalLoss += loss;
                stepCount++;
            }
            catch (NotSupportedException)
            {
                // Native MNN not built with training — re-queue and bail out.
                foreach (var s in samples) await _queue.EnqueueAsync(s, ct).ConfigureAwait(false);
                _logger.LogWarning("[NightlyAdapterTrainer] MNN training not enabled; samples re-queued, run skipped");
                return;
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "[NightlyAdapterTrainer] step failed for sample at {At}", sample.AtUtc);
            }
        }

        if (stepCount > 0)
        {
            try
            {
                _adapter.SaveAdapter(_opts.AdapterPath);
                _adapter.Apply(_opts.AdapterPath);
                _logger.LogInformation(
                    "[NightlyAdapterTrainer] trained {Steps} steps, avg-loss={Loss:F4}, adapter saved to {Path}",
                    stepCount, totalLoss / stepCount, _opts.AdapterPath);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "[NightlyAdapterTrainer] adapter save/apply failed");
            }
        }
    }

    /// <summary>(Phase D3) Char-level tokenizer fallback — every char becomes its UTF-16 code-unit value.</summary>
    private static int[] CharTokenizer(string text)
    {
        if (string.IsNullOrEmpty(text)) return Array.Empty<int>();
        var arr = new int[text.Length];
        for (var i = 0; i < text.Length; i++) arr[i] = text[i];
        return arr;
    }
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/inference/NightlyAdapterTrainer.kt
// NightlyAdapterTrainer.kt
//
// Kotlin port of CircleAI.Inference.NightlyAdapterTrainer (Phase D3). C# is the
// EXACT spec. Periodically drains the FeedbackTrainingQueue, runs LoRA gradient
// steps against the current model handle, saves the adapter to disk, and
// applies it atomically. Idle-and-charging gating is host-supplied via the
// shouldFireNow predicate.
//
// The native LoRAAdapterManager (TrainStep / SaveAdapter / Apply) is injected
// behind ILoRAAdapterManager. A build without native training throws
// TrainingNotSupportedException from trainStep — the trainer re-queues the
// drained batch and bails, exactly like the C# NotSupportedException path.

package com.bhengubv.circleai.inference

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.File
import kotlin.coroutines.coroutineContext

/** Thrown by [ILoRAAdapterManager.trainStep] when native training isn't built. */
class TrainingNotSupportedException(message: String) : RuntimeException(message)

/**
 * The LoRA-adapter native seam. In C# this is the concrete `LoRAAdapterManager`
 * that P/Invokes MNN's training primitives. Injected here so the trainer's
 * drain/train/save/apply loop is testable without a native runtime.
 */
interface ILoRAAdapterManager {
    /**
     * Run one gradient step against [input]/[target] token ids and return the
     * scalar loss. Throws [TrainingNotSupportedException] when the native build
     * lacks training support.
     */
    fun trainStep(input: IntArray, target: IntArray, learningRate: Float, loRARank: Int): Float

    /** Persist the trained adapter to [path]. */
    fun saveAdapter(path: String)

    /** Apply a saved adapter from [path] to the live model handle atomically. */
    fun apply(path: String)
}

/**
 * A working in-memory [ILoRAAdapterManager]. Maintains an accumulating scalar
 * "weight" updated by each step (SGD on a toy squared-error objective) so the
 * loss is real and decreasing, and records save/apply calls. No native runtime.
 */
class InMemoryLoRAAdapterManager(
    private val supportsTraining: Boolean = true,
) : ILoRAAdapterManager {

    /** Running adapter weight (toy scalar parameter). */
    var weight: Double = 0.0
        private set

    /** Count of successful [trainStep] calls. */
    var stepCount: Int = 0
        private set

    /** Last path passed to [saveAdapter]. */
    var lastSavedPath: String? = null
        private set

    /** Last path passed to [apply]. */
    var lastAppliedPath: String? = null
        private set

    override fun trainStep(input: IntArray, target: IntArray, learningRate: Float, loRARank: Int): Float {
        if (!supportsTraining) {
            throw TrainingNotSupportedException("MNN not built with training support.")
        }
        // Toy objective: drive `weight` toward the mean target token id.
        val goal = if (target.isEmpty()) 0.0 else target.average()
        val error = goal - weight
        weight += learningRate.toDouble() * error
        stepCount++
        val loss = error * error
        return loss.toFloat()
    }

    override fun saveAdapter(path: String) {
        lastSavedPath = path
        // Persist a tiny marker so the file exists on disk (parity with a real save).
        val f = File(path)
        f.parentFile?.mkdirs()
        f.writeText("circleai-lora-adapter\nweight:$weight\nsteps:$stepCount\n")
    }

    override fun apply(path: String) {
        lastAppliedPath = path
    }
}

/**
 * Options for [NightlyAdapterTrainer]. Ports NightlyAdapterTrainerOptions.
 *
 * @param minBatchSize Minimum samples to bother training. Skip otherwise.
 * @param maxSamplesPerRun Cap per nightly run so a backlog can't lock the device.
 * @param learningRate Adam-style LR for the LoRA adapter parameters.
 * @param loRARank Rank of the LoRA decomposition; lower = smaller adapter.
 * @param adapterPath Where to persist the trained adapter file.
 * @param intervalMillis How often to check whether to train. Default 6 hours.
 * @param shouldFireNow Optional gate (battery, charging, idle) — defaults to "always".
 * @param tokenizer Text → int IDs. Falls back to char-level mapping when null.
 */
data class NightlyAdapterTrainerOptions(
    val minBatchSize: Int = 16,
    val maxSamplesPerRun: Int = 256,
    val learningRate: Float = 1e-4f,
    val loRARank: Int = 8,
    val adapterPath: String = "circleai-lora.mnn",
    val intervalMillis: Long = 6L * 60 * 60 * 1000,
    val shouldFireNow: (() -> Boolean)? = null,
    val tokenizer: ((String) -> IntArray)? = null,
)

/**
 * (Phase D3) Drains [IFeedbackTrainingQueue] and trains a LoRA adapter. Runs a
 * background loop between [NightlyAdapterTrainerOptions.intervalMillis] ticks;
 * [runOnceAsync] is public so a host can trigger manually.
 */
class NightlyAdapterTrainer(
    private val queue: IFeedbackTrainingQueue,
    private val adapter: ILoRAAdapterManager,
    private val opts: NightlyAdapterTrainerOptions = NightlyAdapterTrainerOptions(),
) {
    private var scope: CoroutineScope? = null
    private var loop: Job? = null

    /** Start the background training loop. Idempotent. */
    fun start(scope: CoroutineScope) {
        if (loop != null) return
        this.scope = scope
        loop = scope.launch(Dispatchers.Default) { loopAsync() }
    }

    /** Stop the background loop and wait for it to unwind. */
    suspend fun stop() {
        loop?.cancelAndJoin()
        loop = null
        scope = null
    }

    private suspend fun loopAsync() {
        while (coroutineContext.isActive) {
            try {
                if (opts.shouldFireNow == null || opts.shouldFireNow!!.invoke()) {
                    runOnceAsync()
                }
            } catch (_: Exception) {
                // run failed — log-and-continue (matches C# LogWarning)
            }
            try {
                delay(opts.intervalMillis)
            } catch (_: Exception) {
                return
            }
        }
    }

    /**
     * (Phase D3) Drain + train in one pass. Skips when fewer than
     * [NightlyAdapterTrainerOptions.minBatchSize] samples are pending. On
     * [TrainingNotSupportedException] the drained batch is re-queued and the
     * run is abandoned. Saves + applies the adapter only when at least one step
     * succeeded. Returns the number of gradient steps taken.
     */
    suspend fun runOnceAsync(): Int {
        if (queue.pending < opts.minBatchSize) return 0

        val samples = queue.drainAsync(opts.maxSamplesPerRun)
        if (samples.isEmpty()) return 0

        val tokenizer = opts.tokenizer ?: ::charTokenizer
        var totalLoss = 0f
        var stepCount = 0

        for (sample in samples) {
            coroutineContext.ensureActive()
            try {
                val input = tokenizer(sample.userText)
                val target = tokenizer(if (sample.polarity >= 0) sample.preferredText else sample.assistantText)
                if (input.isEmpty() || target.isEmpty()) continue

                val loss = adapter.trainStep(input, target, opts.learningRate, opts.loRARank)
                totalLoss += loss
                stepCount++
            } catch (_: TrainingNotSupportedException) {
                // Native MNN not built with training — re-queue and bail out.
                for (s in samples) queue.enqueueAsync(s)
                return 0
            } catch (_: Exception) {
                // step failed for this sample — skip (matches C# LogWarning)
            }
        }

        if (stepCount > 0) {
            try {
                adapter.saveAdapter(opts.adapterPath)
                adapter.apply(opts.adapterPath)
            } catch (_: Exception) {
                // adapter save/apply failed — log-and-continue
            }
        }
        return stepCount
    }

    private companion object {
        /** Char-level tokenizer fallback — every char becomes its UTF-16 code-unit value. */
        fun charTokenizer(text: String): IntArray {
            if (text.isEmpty()) return IntArray(0)
            return IntArray(text.length) { text[it].code }
        }
    }
}
